Add the pyocd as a runner for stm32F303 and stm32f334 nucleo boards

FRASTM opened the pull request:
Add the pyocd as a runner for the nucleo_f303k8 and nucleo_f334r8 boards for flashing and debugging.

If the stm32f303k8tx/ stm32f334r8tx  is not in the list of targets for pyocd :
`$ pyocd list --targets`
then add the pack for pyocd
`$ pyocd pack install stm32f303`  `$ pyocd pack install stm32f334`  

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54117

Nucleo_f334r8 and nucleo_f303k8 report the same error with openocd : unable to write the flash after erasing

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32f303 nucleo with pyocd and cubeprogrammer as runner

Add the pyocd ant the stm32cubeprogrammer as a runner of the
nucleo_f303k8 platforms,
when the openocd gives an flashing error.
Add the pyocd pack if needed : 'pyocd pack install stm32f303'

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32f334 nucleo with pyocd and cubeprogrammer as runner

Add the pyocd ant the stm32cubeprogrammer as a runner
of the nucleo_f334r8 platforms,
when the openocd gives an flashing error.
Add the pyocd pack if needed : 'pyocd pack install stm32f334

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented:
@moritz-geier can you have a try?

moritz-geier commented:
Since the target stm32f303 is missing I tried to download it, however the packet command `pyocd pack install stm32f303` fails with `0000391 W No matching devices. Please make sure the pack index is up to date. [pack_cmd]` and if i try to run `pyocd pack update` I get
```
0000355 I Updating pack index... [pack_cmd]
08:34:23 [INFO] Downloaded http://www.keil.com/pack/index.pidx
08:34:23 [INFO] Found 833 Pdsc entries
08:34:24 [WARN] No Core found!33)
08:34:24 [WARN] No Core found!
... 44 Times the same message ...
08:34:24 [WARN] No Core found!
08:34:24 [WARN] XML error: Expecting </link> found </head>
08:34:24 [WARN] the end of the document has been reached prematurely
08:34:24 [WARN] Device found without a processor ATSAMDA1E14A
08:34:24 [WARN] Device found without a processor ATSAME70J19
08:34:24 [WARN] Device found without a processor ATSAMS70J19
08:34:24 [WARN] Device found without a processor ATSAMV70J19
08:34:24 [WARN] Device found without a processor ATSAMV71J19
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] Support for two parent processors not implemented!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] Support for two parent processors not implemented!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] XML error: Expecting </P> found </BODY>
... 62 Times the same message ...
08:34:24 [WARN] XML error: Expecting </P> found </BODY>
08:34:25 [WARN] Support for two parent processors not implemented!
```
is this related?

FRASTM commented:
Only if you do not find the stm32f303xx  in the list of your pyocd targets 
(pyocd list --targets or pyocd list --targets | grep stm32f303)  
you should install that pack 
$pyocd pack install stm32f303
should download the pack 
```
Downloading packs (press Control-C to cancel):
    Keil.STM32F3xx_DFP.2.2.2
Downloading descriptors (001/001)
```

moritz-geier commented:
If I call the `pyocd list --targets | grep stm32f303` nothing returns, thus the target is not installed. When I try to install the package as you said I get the error messages described above. I wrote an [Issue on the pyocd page](https://github.com/pyocd/pyOCD/issues/1496) as soon as this is resolved I will test the changes.

FRASTM commented:
Adding the stm32CubeProgrammer as a runner for west flash

FRASTM committed:
soc: arm: stm32f3 enable Debug Module during SLEEP mode

This patch allows successive reflashing operation on stm32f3
boards by Enabling the Debug Module during SLEEP mode.
This will especially makes reflahing and debugging possible
with pyocd runner on west commands.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
> If I call the `pyocd list --targets | grep stm32f303` nothing returns, thus the target is not installed. When I try to install the package as you said I get the error messages described above. I wrote an [Issue on the pyocd page](https://github.com/pyocd/pyOCD/issues/1496) as soon as this is resolved I will test the changes.

For flashing you can still use the stm32cubeprogrammer, but for debugging the pyocd target pack is required.

moritz-geier commented:
Got it running, I'm now able to flash and debug via pyocd.

tagunil commented on soc/arm/st_stm32/stm32f3/soc.c:
```diff
@@ -43,6 +44,9 @@ static int stm32f3_init(const struct device *arg)
 	SystemCoreClock = 8000000;
 
+	/* Allow reflashing the board */
+	LL_DBGMCU_EnableDBGSleepMode();
```
What about other STM32 series, wouldn't this change be useful as well?
